File: MyLeetCodeAnswers/C++/pathSumII.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

/*Improved*/
class Solution {
public:
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        vector<vector<int> > result;
        if (root == NULL) return result;
        vector<int> path;
        pathSum(result, path, sum, root);
        return result;
    }
    
    void pathSum(vector<vector<int> > &result, vector<int> &path, int sum, TreeNode *root) {
        path.push_back(root->val);
        if (root->left == NULL && root->right == NULL && root->val == sum) {
            result.push_back(path);
        } else {
            if (root->left != NULL)
                pathSum(result, path, sum-root->val, root->left);
            if (root->right != NULL)
                pathSum(result, path, sum-root->val, root->right);
        }
        path.pop_back();
    }
};

/*My recursive solution*/
class Solution {
public:
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        vector<vector<int> > result;
        if (root == NULL) return result;
        vector<int> path;
        int curSum = 0;
        pathSum(result, path, sum, curSum, root);
        return result;
    }
    
    void pathSum(vector<vector<int> > &result, vector<int> &path, int sum, int &curSum, TreeNode *root) {
        path.push_back(root->val);
        curSum+= root->val;
        if (root->left == NULL && root->right == NULL) {
            if (curSum == sum) {
                result.push_back(path);
            }
        } else {
            if (root->left != NULL)
                pathSum(result, path, sum, curSum, root->left);
            if (root->right != NULL)
                pathSum(result, path, sum, curSum, root->right);
        }
        curSum -= root->val;
        path.pop_back();
    }
};
int main() {
    return 0;
}

File: MyLeetCodeAnswers/java/PathSumII.java
import java.util.ArrayList;
import java.util.List;

/**
 * Source   : https://leetcode.com/problems/path-sum-ii/
 * Author   : Wang Meng
 * Date     : 2016-03-12
 * Given a binary tree and a sum, find all root-to-leaf paths where each path's sum equals the given sum.
 * <p>
 * For example:
 * Given the below binary tree and sum = 22,
 * return
 * [
 * [5,4,11,2],
 * [5,8,4,5]
 * ]
 * 5
 * / \
 * 4   8
 * /   / \
 * 11  13  4
 * /  \    / \
 * 7   2   5  1
 */
public class PathSumII {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> ret = new ArrayList<>();
        List<Integer> cur = new ArrayList<>();
        pathSum(root, sum, cur, ret);
        return ret;
    }

    private void pathSum(TreeNode root, int sum, List<Integer> cur, List<List<Integer>> ret) {
        if (root == null) {
            return;
        }
        cur.add(root.val);
        if (root.left == null && root.right == null && root.val == sum) {
            ret.add(new ArrayList<>(cur));
        } else {
            pathSum(root.left, sum - root.val, cur, ret);
            pathSum(root.right, sum - root.val, cur, ret);
        }
        cur.remove(cur.size() - 1);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(5);
        root.left = new TreeNode(4);
        root.left.left = new TreeNode(11);
        root.left.left.left = new TreeNode(7);
        root.left.left.right = new TreeNode(2);
        root.right = new TreeNode(8);
        root.right.left = new TreeNode(13);
        root.right.right = new TreeNode(4);
        root.right.right.left = new TreeNode(5);
        root.right.right.right = new TreeNode(1);
        System.out.println(new PathSumII().pathSum(root, 22));
    }
}
